File: ruby/src/strings/anagram.rb
#check if 2 strings are anagram or not - re-arranging can result in same string or not

#Method 1: str1.chars.sort.join == str2.chars.sort.join (Sort and compare)
def anagram?(str1, str2)
  return false if str1.size != str2.size
  hash = {}
  for i in 0..str1.size-1
    hash[str1[i]] ||= 0
    hash[str1[i]] += 1
  end

  for i in 0..str2.size-1
    val = hash[str2[i]]
    return false if val.nil? || val <= 0
    hash[str2[i]] -= 1
  end

  !hash.find {|k,v|v!=0}
end
puts anagram?("ER","RE")
puts anagram?("QER","RE")
puts anagram?("QER","RES")
puts anagram?("QER","QER")
puts anagram?("QERR","QERQ")

File: java/src/string/Anagram.java
package src.string;

import java.util.HashMap;
import java.util.Map;

/*
 * Check if 2 strings are anagram or not - re-arranging can result in same string or not
 */
public class Anagram {
	public boolean isAnagram(String str1, String str2) {
		// Method 1: Sort and compare. O(n log n) for sorting
		// Method 2: count number of chars in hash map and compare : O(n)
		
		if (str1 == null || str2 == null || str1.length() != str2.length()) { // Length must be same for anagrams
			return false;
		}
		
		Map<Character, Integer> map = new HashMap<Character, Integer>();
		
		//for (char ch: str1.toCharArray()) {
		for (int i=0; i<str1.length(); i++) {
			char ch = str1.charAt(i);
			int count = 1;
			if (map.containsKey(ch)) {
				count += map.get(ch);
			}
			map.put(ch, count);
		}
		int count;
		//for (char ch: str2.toCharArray()) {
		for (int i=0; i<str2.length(); i++) {
				char ch = str2.charAt(i);
			if (!map.containsKey(ch) || ((count = map.get(ch)) <= 0)) {
				return false;
			}
			if (count > 1) {
				map.put(ch, count-1);
			} else {
				map.remove(ch);
			}
		}
		return map.size() == 0;
	}

	public static void main(String[] args) {
		System.out.println("Expect true: " + new Anagram().isAnagram("Anagram", "Anagamr"));
		System.out.println("Expect false: " + new Anagram().isAnagram("Anar", "Anag"));
		System.out.println("Expect false: " + new Anagram().isAnagram("Anag", "Anagr"));
	}
}
